### PR TITLE
kubeadm: modify SetKubernetesVersion

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:

Without this patch SetKubernetesVersion does not set
a version in all locations where it's called, because
the passed config to it always has a default version
stable-1.

This always triggers the != "" check and the function
returns without setting a version.

Validate against DefaultKubernetesVersion instead.

This fixes all cases where fetching a version from the internet
is not needed at all - e.g. "kubeadm token create".

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
refs #https://github.com/kubernetes/kubeadm/issues/1166

**Special notes for your reviewer**:
NONE

**Release note**:

```release-note
kubeadm: fix a case where fetching a kubernetesVersion from the internet still happened even if some commands don't need it.
```

/assign @xiangpengzhao 
/assign @fabriziopandini 
/cc @timothysc 
/kind bug

Peter, this is related to your patch about:
```
// KubernetesVersion is not used, but we set it explicitly to avoid the lookup
// of the version from the internet when executing ConfigFileAndDefaultsToInternalConfig
```
https://github.com/kubernetes/kubernetes/commit/5cf9291e025baf3efa13e8ef73dcfa7fb46ee925

WDYT?
i don't see a problem with having the function like that in the current codebase - i.e it would override with the client version, but this will avoid fetching from the internet if labels are passed.

also, it's not a critical bug, but i'd vote for us to cherry pick this for 1.12.

### CODE REVIEW

**Comment by neolit123** [comment]
found issue with the earlier patch, applied modification and renamed the PR too.

File: cmd/kubeadm/app/cmd/phases/util.go [hunk 1]
```diff
@@ -65,9 +65,11 @@ func runCmdPhase(cmdFunc func(outDir string, cfg *kubeadmapi.InitConfiguration)
 	}
 }
 
-// SetKubernetesVersion gets current Kubeadm version and sets it as Kubernetes version of master configuration if not set.
+// SetKubernetesVersion gets the current Kubeadm version and sets it as KubeadmVersion in the config,
+// unless it's already set to a value different from the default.
 func SetKubernetesVersion(client clientset.Interface, cfg *kubeadmapiv1beta1.InitConfiguration) error {
-	if cfg.KubernetesVersion != "" {
+
+	if cfg.KubernetesVersion != kubeadmapiv1beta1.DefaultKubernetesVersion {
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/cmd/phases/util.go]: IIUC, the func will return from here if `cfg.KubernetesVersion == ""`. This is not what we expect, right?

**Review by xiangpengzhao — COMMENTED** [review]
@neolit123 Agree with your idea :+1: EDIT: I misread the PR body. You're correct on the condition change.

**Review by xiangpengzhao — COMMENTED** [review]
Just one concern, otherwise LGTM. Thanks for fixing this! @neolit123

File: cmd/kubeadm/app/cmd/phases/util.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bart0sh) [reply cmd/kubeadm/app/cmd/phases/util.go]: /lgtm

[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/util.go]: @xiangpengzhao 

hm, so if the default changes to `""` the branch will handle it and return.
but i'm wondering if we should handle `""` with the current default not being `""`.

```
if cfg.KubernetesVersion != kubeadmapiv1beta1.DefaultKubernetesVersion && cfg.KubernetesVersion != "") {
    return nil
}
```

^ i  assume this is what we want?

**Review by timothysc — CHANGES_REQUESTED** [review]
This points to a gap in testing, could you please add a unit test for the different scenarios.

File: cmd/kubeadm/app/cmd/phases/util.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/util.go]: sounds good.
will update the PR tomorrow.

[the remark this answers is not in the document]
Comment (xiangpengzhao) [reply cmd/kubeadm/app/cmd/phases/util.go]: SGTM :)

**Comment by neolit123** [comment]
updated:
- added a unit test for SetKubernetesVersion
- modified the function to use version.Get().String() directly too.

File: cmd/kubeadm/app/cmd/phases/util_test.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+			input:  "",
+			output: ver,
+		},
+		{
+			name:   "default version is skipped",
+			input:  kubeadmapiv1beta1.DefaultKubernetesVersion,
+			output: ver,
+		},
+		{
+			name:   "any other version is processed",
```
Comment (xiangpengzhao) [inline cmd/kubeadm/app/cmd/phases/util_test.go]: Seems like the logic of `SetKubernetesVersion` is "empty version and default version is processed and any other version is skipped". @neolit123 WDYT?

**Review by xiangpengzhao — COMMENTED** [review]
Good refactoring :+1: Just one concern, otherwise LGTM.

File: cmd/kubeadm/app/cmd/phases/util_test.go (hunk 2 above)
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/util_test.go]: sorry, was in a rush and got things mixed up. i will update the PR later today.

**Review by timothysc — APPROVED** [review]
Thanks for the cleanup @neolit123 

/lgtm 
/approve

**Pull request merged**